File: app/src/main/java/cn/tongdun/android/adapter/BaseRecyclerViewAdapter.java
package cn.tongdun.android.adapter;

import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

/**
 * @description: RecyclerView Adapter
 * @author: wuzuchang
 * @date: 2022/12/6
 */
public abstract class BaseRecyclerViewAdapter<T> extends RecyclerView.Adapter<BaseRecyclerViewAdapter.BaseViewHolder> {

    public List<T> mData;

    public BaseRecyclerViewAdapter(List<T> data) {
        this.mData = data;
    }

    public abstract int getLayoutId(int viewType);

    @Override
    public BaseViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return BaseViewHolder.get(parent, getLayoutId(viewType));
    }

    @Override
    public void onBindViewHolder(@NonNull BaseViewHolder holder, int position) {
        convert(holder, mData.get(position), position);
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    public abstract void convert(BaseViewHolder holder, T data, int position);

    static class BaseViewHolder extends RecyclerView.ViewHolder {
        private SparseArray<View> mViews;
        private View mConvertView;

        private BaseViewHolder(View v) {
            super(v);
            mConvertView = v;
            mViews = new SparseArray<>();
        }

        public static BaseViewHolder get(ViewGroup parent, int layoutId) {
            View convertView = LayoutInflater.from(parent.getContext()).inflate(layoutId, parent, false);
            return new BaseViewHolder(convertView);
        }

        public <T extends View> T getView(int id) {
            View v = mViews.get(id);
            if (v == null) {
                v = mConvertView.findViewById(id);
                mViews.put(id, v);
            }
            return (T) v;
        }

        public void setText(int id, String value) {
            TextView view = getView(id);
            view.setText(value);
        }
    }
}

File: app/src/main/kotlin/cn/tongdun/android/adapters/BaseRecyclerViewAdapter.kt
package cn.tongdun.android.adapters

import android.util.SparseArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

/**
 * @description:
 * @author: wuzuchang
 * @date: 2023/5/16
 */
abstract class BaseRecyclerViewAdapter<T>(val data: List<T>) : RecyclerView.Adapter<BaseRecyclerViewAdapter.BaseViewHolder>() {


    abstract fun getLayoutId(viewType: Int): Int

    abstract fun convert(holder: BaseViewHolder, data: T, position: Int)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        return BaseViewHolder[parent, getLayoutId(viewType)]
    }

    override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
        convert(holder, data[position], position)
    }

    override fun getItemCount(): Int {
        return data.size
    }

    class BaseViewHolder private constructor(private val mConvertView: View) :
        RecyclerView.ViewHolder(
            mConvertView
        ) {
        private val mViews: SparseArray<View> = SparseArray()

        fun <T : View> getView(id: Int): T? {
            var v = mViews[id]
            if (v == null) {
                v = mConvertView.findViewById(id)
                mViews.put(id, v)
            }
            return v as T?
        }

        fun setText(id: Int, value: String) {
            val view = getView<TextView>(id)
            view?.text = value
        }

        companion object {
            operator fun get(parent: ViewGroup, layoutId: Int): BaseViewHolder {
                val convertView =
                    LayoutInflater.from(parent.context).inflate(layoutId, parent, false)
                return BaseViewHolder(convertView)
            }
        }
    }
}
